learn/hoon/hoon-tutorial/egg-timer.md

```diff
@@ -0,0 +1,120 @@
++++
+title = "2.7.1 Walkthrough: Egg Timer"
+weight = 36
+template = "doc.html"
++++
+
+The Arvo operating system is divided up into modules called vanes. Gall is the vane responsible for providing an API for building stateful applications in Urbit. Gall is a complex system that takes a lot of getting used to, so let's dip our toes in with a simple egg-timer app.
+
+
+```
+|%
++$  effect   (pair bone syscall)
++$  syscall  [%wait path @da]
+--
+|_  [bowl:gall ~]
+++  poke-noun
+  |=  t=@dr
+  ^+  [*(list effect) +>.$]
+  :_  +>.$  :_  ~
+  [ost %wait /(scot %da now) (add now t)]
+++  wake
+  |=  [=wire error=(unit tang)]
+  ^+  [*(list effect) +>.$]
+  ~&  "Timer went off!"
+  [~ +>.$]
+--
+```
+
+The first thing to notice is that we are creating a `core` (`|%`) and a `door` (`|_`). This is a typical style of Gall programming where your types are defined in the first `core` and your application is defined in the following `door`.
+
+```
+|%
++$  effect   (pair bone syscall)
++$  syscall  [%wait path @da]
+--
+```
+
+The `core` here defines two types: `effect` and `syscall`. An `effect` is a `pair` of `bone` and `syscall`. A `bone` is a Gall-only type that identifies app event chains by mapping atoms to them. We define `syscall` to be a request to Arvo to do something for us. In this case, the only valid `syscall` will be `%wait` which we'll discuss in a bit.
+
+It's important to note that the names `effect` and `syscall` are arbitrary; you can call them whatever you'd like. Commonly you will see them called `move` and `card`, respectively, but we've changed the names here to try to introduce some clarity that that older convention lacks.
+
+The sample of the `door` is:
+
+```
+[bowl:gall ~]
+```
+
+You can find the full definition of `bowl` in `sys/zuse.hoon`, but for now it's enough to know that this is the default app state and includes various faces for information. Below are some important such faces:
+
+`our`  The current ship
+`src`  The ship the request originated with
+`ost`  A reference to a chain of events
+`eny`  Entropy
+`now`  The current time of the event
+
+The `door` we've made has two arms `poke-noun` and `wake`. Gall is capable of dispatching `pokes`, or requests, to an app based on the mark of the data given along with that poke. These are sent to the arm with the name that matches the mark of the data. Here we use the generic `noun` mark:
+
+```
+++  poke-noun
+  |=  t=@dr
+  ^+  [*(list effect) +>.$]
+  :_  +>.$  :_  ~
+  [ost %wait /(scot %da now) (add now t)]
```

Comment:
Why `/(scot %da now)` instead of a less opaque wire like `/egg-timer`? This (theoretically, I wouldn't encourage it for this tutorial per se) allows you to do `++wake-egg-timer`, or at least have a more readable `wire` in `++wake`. Using _just_ a plain, "random" timestamp for the wire seems ill-advised.

Reply:
given that we're not dispatching on it, i'm not sure that it matters?  but probably yeah.

Reply:
It does indeed not matter technically, but it's semantically incorrect and kinda bad practice.

Reply:
Fair enough. If you use /egg-timer though don't you have a potential problem with say, setting two timers at once for different times?

Reply:
They'll still work if that's what you're asking. Duplicate wires are fine iirc.